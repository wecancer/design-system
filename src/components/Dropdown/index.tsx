import React, {useState, useRef} from 'react'
import styled, {css} from 'styled-components'

import OutsideEvent from '../OutsideEvent'

type AxisX = 'left' | 'right'

const Wrapper = styled(OutsideEvent)`
  position: relative;
  display: inline-block;
`

const TriggerContainer = styled.div``

const BoxContainer = styled.div<{triggerHeight: number; axisX: AxisX}>`
  ${({triggerHeight, axisX}) => css`
    position: absolute;
    top: ${triggerHeight + 10}px;
    ${axisX === 'left'
      ? css`
          left: 0;
        `
      : css`
          right: 0;
        `}
  `}
`

export type ActionParams = {
  isOpen: boolean
  handleClose(): void
  handleToggle(): void
}

export type Props = {
  axisX?: AxisX
  trigger(params: ActionParams): React.ReactNode
  children(params: ActionParams): React.ReactNode
  className?: string
}

const Dropdown = ({axisX = 'left', children, trigger, ...props}: Props) => {
  const refTrigger = useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = useState(false)
  const handleClose = () => setOpen(false)
  const handleToggle = () => setOpen(!isOpen)

  return (
    <Wrapper onClickOutside={handleClose} {...props}>
      <TriggerContainer ref={refTrigger}>
        {trigger({isOpen, handleClose, handleToggle})}
      </TriggerContainer>
      {isOpen && (
        <BoxContainer
          axisX={axisX}
          triggerHeight={refTrigger.current?.offsetHeight || 0}
        >
          {children({isOpen, handleClose, handleToggle})}
        </BoxContainer>
      )}
    </Wrapper>
  )
}

export default Dropdown
