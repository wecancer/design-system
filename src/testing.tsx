import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
