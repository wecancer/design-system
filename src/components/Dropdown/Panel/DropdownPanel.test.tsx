import Dropdown from '.'
import {render, fireEvent} from '../../../testing'

describe('<DropdownPanel />', () => {
  const options = [
    {
      label: 'item 1',
      value: '01',
    },
    {
      label: 'item 2',
      value: '02',
    },
    {
      label: 'item 3',
      value: '03',
    },
  ]

  it('should render snapshot correctly', () => {
    const {container} = render(
      <>
        <Dropdown label="empty options" onChange={() => {}} options={[]} />
        <Dropdown label="with " onChange={() => {}} options={options} />
        <Dropdown
          label="testing"
          onChange={() => {}}
          options={options}
          value={options[0]}
        />
      </>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('must trigger an event of change when the element has been selected', () => {
    const index = 0
    const option = options[index]
    const {queryByText, queryByRole} = render(
      <Dropdown
        label="Click me!"
        onChange={(value) => expect(option).toMatchObject(value.option)}
        options={options}
      />,
    )
    const triggerButton = queryByText('Click me!')
    fireEvent.click(triggerButton as HTMLElement)

    const buttonSelect = queryByRole('listbox')?.querySelectorAll(
      '[role="button"]',
    )[index] as HTMLElement
    fireEvent.click(buttonSelect)
  })

  it('should be use the label of value', () => {
    const option = options[0]
    const {queryByText} = render(
      <Dropdown
        value={option}
        label="Click me!"
        options={options}
        onChange={(value) => expect(option).toMatchObject(value.option)}
      />,
    )

    expect(queryByText('Click me!')).toBeNull()
    expect(queryByText(option.label)).toBeDefined()
  })
})
