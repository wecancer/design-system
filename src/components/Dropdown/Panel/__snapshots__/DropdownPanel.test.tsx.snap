// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownPanel /> should render snapshot correctly 1`] = `
.c0 {
  position: relative;
  display: inline-block;
}

.c1 {
  cursor: pointer;
  font-family: Poppins,sans-serif;
}

.c2 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  margin: 0;
}

<div
  class="c0"
>
  <div
    class=""
  >
    <div
      class="c1"
      role="button"
      tabindex="0"
    >
      <p
        class="c2"
      >
        empty options
      </p>
    </div>
  </div>
</div>
`;
